src/index.ts

```diff
@@ -1,19 +1,649 @@
+
+// Type exports:
+
+export {
+  DispatchResult,
+  SignMessageOptions,
+} from "./types/arconnect/arconnect.types";
+export { TypedArray, BinaryDataType } from "./lib/utils/arweaveUtils";
+
+export {
+  Auth0Strategy,
+  IdTokenWithData,
+  UserDetails,
+} from "./lib/auth/auth0.types";
+
+// Constant exports:
+
+export {
+  DEFAULT_OTHENT_CONFIG,
+  CLIENT_NAME,
+  CLIENT_VERSION,
+} from "./lib/config/config.constants";
+
+// TODO: Polyfill. Should we overwrite a global like this from a library?
+
 window.Buffer = Buffer;
+
+export interface OthentConfig {
+  auth0Domain: string;
+  auth0ClientId: string;
+  auth0Strategy: Auth0Strategy;
+  serverBaseURL: string;
+}
+
+export interface OthentOptions extends Partial<OthentConfig> {
+  crypto?: Crypto | null;
+}
+
+export class Othent
+  implements Omit<ArConnect, "connect" | "dispatch" | "signDataItem">
+{
+  walletName = CLIENT_NAME;
+
+  walletVersion = CLIENT_VERSION;
+
+  config: OthentConfig = DEFAULT_OTHENT_CONFIG;
+
+  crypto: Crypto;
+
+  api: OthentKMSClient;
+
+  auth0: OthentAuth0Client;
+
+  // TODO: Add listener for user details change?
+
+  // TODO: Add listener for errors and a silentErrors: boolean property?
+
+  // TODO: When using refresh tokens in memory, the developer has to manually call connect() before calling any other function, as otherwise
+  // get a "Missing cached user." error. Can we improve that?
+
+  constructor(options: OthentOptions = {}) {
+    let { crypto: cryptoOption, ...configOptions } = options;
+
+    this.config = { ...configOptions, ...DEFAULT_OTHENT_CONFIG };
+
+    let crypto = cryptoOption;
+
+    if (!crypto) {
+      if (typeof window !== "undefined") {
+        crypto = window.crypto;
+      } else if (typeof global !== "undefined") {
+        crypto = global.crypto;
+      } else if (crypto === undefined) {
+        throw new Error(
+          "A Crypto module is needed to use `signMessage` and `verifyMessage`. If you are sure you won't be using those, pass `crypto: null` as an option.",
+        );
+      }
+    }
+
+    this.crypto = crypto!;
+
+    this.auth0 = new OthentAuth0Client(
+      this.config.auth0Domain,
+      this.config.auth0ClientId,
+      this.config.auth0Strategy,
+    );
+
+    this.api = new OthentKMSClient(this.config.serverBaseURL, this.auth0);
+
+    if (process.env.NODE_ENV === "development") {
+      console.log(process.env);
+
+      console.log(`${this.walletName} @ ${this.walletVersion}`);
+
+      Object.entries(this.config).map(([key, value]) => {
+        console.log(` ${key.padStart(13)} = ${value}`);
+      });
+    }
+  }
+
+  async init() {
+    await this.auth0.init();
+  }
+
+  // CONNECT / DISCONNECT:
+
+  /**
+   * Connect the users account, this is the same as login/signup in one function.
+   * @returns The the users details.
+   */
+  async connect(): Promise<UserDetails | null> {
+    // Call `getTokenSilently()` to reconnect if we still have a valid token / session.
```

Comment:
Connect flow is commented step by step as, when using a custom Auth0 client and or server, here's where most of the issues will happen.

Reply:
🙌